### PR TITLE
[SYCL] Switch SPIR-V offload target to opaque pointers

### PR DESCRIPTION
**Pull request by bader**
Although there are a few tests failing due to this change, we need to go with this change to avoid future regressions and unblock changes removing typed pointers support. The regressions are supposed to be fixed by follow-up patches.

### CODE REVIEW

File: sycl/test-e2e/KernelFusion/internalize_array_wrapper.cpp
Comment (bader) [inline sycl/test-e2e/KernelFusion/internalize_array_wrapper.cpp:7]: FYI: compiler gives me a warning that -fsycl-embed-ir is ignored.
Kernel fusion developers might need to check if this option is needed for test.

**Review by v-klochkov — COMMENTED** [review]
I don't feel comfortable doing the change that breaks one of the most basic functionality such as sin/cos/pow, etc.
It may make ESIMD not usable for a whole category of users.

@sarnex - Is that the test that you mentioned once when said: "There is one test failing with opaque pointers and the issue is in GPU RT" ?
If Yes, do you know what case is not handled correctly in GPU RT, does it affect all math functions and is there a workaround?
@turinevgeny - FYI.

**Comment by sarnex** [comment]
@v-klochkov 

> Is that the test that you mentioned once when said: "There is one test failing with opaque pointers and the issue is in GPU RT" ?

Yes, ext_math is the test in question, and the issue was in the GPU RT. It should be fixed with https://github.com/intel/intel-graphics-compiler/commit/c083decc550cbfcd5bf5412b75acba7b14ee2ac2 which made it into igc 1.0.14342. The CI here is a bit confusing so I don't know what igc version is being used here, but I don't think we are using that version even with internal testing.

@bader Can you please help me find the igc version being used in CI here?

**Comment by bader** [comment]
> Can you please help me find the igc version being used in CI here?

Linux - https://github.com/intel/llvm/blob/sycl/devops/dependencies.json#L9-L14.
Windows version is installed on runners directly by @intel/dpcpp-devops-reviewers team, so I think they can tell if the information from https://github.com/intel/llvm/blob/sycl/devops/dependencies.json#L50-L54 is relevant.

**Comment by sarnex** [comment]
Thanks, so definitely the driver is just too old.

Let me investigate with that driver version and see what the scope is and if I can find any workarounds.

**Comment by sarnex** [comment]
@v-klochkov 
It looks like this is pretty severe and fundamental with no workarounds I could find. Even the below reproduces the issue:

```
cgh.parallel_for<class FillBuffer>(
        NumOfWorkItems, [=](sycl::id<1> WIid) SYCL_ESIMD_KERNEL {
sycl::sin<8>(0);
});
```

So any use of the SYCL math functions in a ESIMD kernel seems to cause the problem.

File: sycl/test-e2e/KernelFusion/internalize_array_wrapper.cpp
Comment (sommerlukas) [reply sycl/test-e2e/KernelFusion/internalize_array_wrapper.cpp:7]: The option is only needed on non-SPIR-V targets (currently only CUDA), so it's fine if the option is ignored when compiling for SPIR-V. 

I assume the test only failed when running it, not during compilation, right?

**Comment by jsji** [comment]
> So any use of the SYCL vector math functions in a ESIMD kernel seems to cause the problem.

> Yes, ext_math is the test in question, and the issue was in the GPU RT. It should be fixed with https://github.com/intel/intel-graphics-compiler/commit/c083decc550cbfcd5bf5412b75acba7b14ee2ac2 which made it into igc 1.0.14342.


@sarnex Is this problem happen only with the *OLD* driver? Or updating the igc will fix these?

**Comment by sarnex** [comment]
@jsji A driver with IGC 1.0.14342 will not have the problem and the test will pass.

File: sycl/test-e2e/KernelFusion/internalize_array_wrapper.cpp
Comment (bader) [reply sycl/test-e2e/KernelFusion/internalize_array_wrapper.cpp:7]: > I assume the test only failed when running it, not during compilation, right?

@sommerlukas here is the error message from the [pre-commit logs](https://github.com/intel/llvm/actions/runs/5731980488/job/15535109389).

```
FAIL: SYCL :: KernelFusion/internalize_array_wrapper.cpp (1038 of 1567)
******************** TEST 'SYCL :: KernelFusion/internalize_array_wrapper.cpp' FAILED ********************
Script:
--
: 'RUN: at line 2';    /__w/llvm/llvm/toolchain/bin//clang++   -fsycl -fsycl-targets=spir64 /__w/llvm/llvm/llvm/sycl/test-e2e/KernelFusion/internalize_array_wrapper.cpp -Xclang -no-opaque-pointers -fsycl-embed-ir -O2 -o /__w/llvm/llvm/build-e2e/KernelFusion/Output/internalize_array_wrapper.cpp.tmp.out
: 'RUN: at line 3';   env ONEAPI_DEVICE_SELECTOR=ext_oneapi_level_zero:gpu  /__w/llvm/llvm/build-e2e/KernelFusion/Output/internalize_array_wrapper.cpp.tmp.out
: 'RUN: at line 3';   env ONEAPI_DEVICE_SELECTOR=opencl:gpu  /__w/llvm/llvm/build-e2e/KernelFusion/Output/internalize_array_wrapper.cpp.tmp.out
: 'RUN: at line 3';   env ONEAPI_DEVICE_SELECTOR=opencl:cpu  /__w/llvm/llvm/build-e2e/KernelFusion/Output/internalize_array_wrapper.cpp.tmp.out
--
Exit Code: 1

Command Output (stdout):
--
$ ":" "RUN: at line 2"
note: command had no output on stdout or stderr
$ "/__w/llvm/llvm/toolchain/bin//clang++" "-fsycl" "-fsycl-targets=spir64" "/__w/llvm/llvm/llvm/sycl/test-e2e/KernelFusion/internalize_array_wrapper.cpp" "-Xclang" "-no-opaque-pointers" "-fsycl-embed-ir" "-O2" "-o" "/__w/llvm/llvm/build-e2e/KernelFusion/Output/internalize_array_wrapper.cpp.tmp.out"
# command stderr:
clang++: warning: argument unused during compilation: '-fsycl-embed-ir' [-Wunused-command-line-argument]
/__w/llvm/llvm/toolchain/bin/llvm-link: /tmp/lit-tmp-fznnrqd4/libsycl-crt-8cab33.o: error: Opaque pointers are only supported in -opaque-pointers mode (Producer: 'LLVM17.0.0git' Reader: 'LLVM 17.0.0git')
/__w/llvm/llvm/toolchain/bin/llvm-link: error:  loading file '/tmp/lit-tmp-fznnrqd4/libsycl-crt-8cab33.o'
clang++: error: sycl-link command failed with exit code 1 (use -v to see invocation)

error: command failed with exit status: 1

--
```

It looks like JIT compiler for fusion component still uses typed pointers. Am I right?

**Comment by sarnex** [comment]
@jsji Sounds good, let me know if you hit any problems with this test on a new driver and I will investigate.

**Comment by jsji** [comment]
@bader @aelovikov-intel @dpcpp-devops-reviewers Can we try to update the IGC version?

@v-klochkov  If we can't update the IGC version due to some other reason for now, can we XFAIL the test as @sarnex confirmed that this is just due to old driver. We will eventually update the IGC version anyhow.

**Comment by sarnex** [comment]
@jsji The problem is not really about this test in particular, the problem is basic ESIMD functionality will break for people not using the latest drivers. Even in internal testing we don't seem to be using a driver that has the fix.

File: sycl/test-e2e/KernelFusion/internalize_array_wrapper.cpp
Comment (sommerlukas) [reply sycl/test-e2e/KernelFusion/internalize_array_wrapper.cpp:7]: In that case I would expect most kernel fusion related tests to fail, not just this one. 

Also, the error is from `sycl-link`, so already seems to happen at compile time, whereas errors in the JIT compiler would only show up at runtime of the application. 

Is the problem maybe that `"-Xclang" "-no-opaque-pointers" ` is still passed in the compilation of the test?

**Comment by bader** [comment]
> @bader @aelovikov-intel @intel/dpcpp-devops-reviewers Can we try to update the IGC version?

I have no objections to check, but we need to test if using custom IGC version doesn't break other tests. We use the version specified by the GPU driver - https://github.com/intel/compute-runtime/releases, which is tested with the rest of the driver components.
@aelovikov-intel, could you run as much tests as possible with the IGC 1.0.14342 or later before the switch? I used to use regular pre-commit, but AFAIK, it doesn't work for testing changes in CI anymore.

File: sycl/test-e2e/KernelFusion/internalize_array_wrapper.cpp
Comment (bader) [reply sycl/test-e2e/KernelFusion/internalize_array_wrapper.cpp:7]: > Is the problem maybe that "-Xclang" "-no-opaque-pointers" is still passed in the compilation of the test?

I don't see these options in the command line. Why do you think they are still passed?

Comment (sommerlukas) [reply sycl/test-e2e/KernelFusion/internalize_array_wrapper.cpp:7]: Because of this line from the error log you posted: 
```
"/__w/llvm/llvm/toolchain/bin//clang++" "-fsycl" "-fsycl-targets=spir64" "/__w/llvm/llvm/llvm/sycl/test-e2e/KernelFusion/internalize_array_wrapper.cpp" "-Xclang" "-no-opaque-pointers" "-fsycl-embed-ir" "-O2" "-o" "/__w/llvm/llvm/build-e2e/KernelFusion/Output/internalize_array_wrapper.cpp.tmp.out"
```

Comment (sommerlukas) [reply sycl/test-e2e/KernelFusion/internalize_array_wrapper.cpp:7]: Which corresponds to the first `RUN` line in the log.

**Comment by jsji** [comment]
> Don't we have ESIMD tests checking "basic functionality"? I would recommend carefully reviewing ESIMD tests as from my POV the ROI of current tests is low.

Yes, agree! But I think this shouldn't be a blocker for opaque pointer enablement.

File: sycl/test-e2e/KernelFusion/internalize_array_wrapper.cpp
Comment (bader) [reply sycl/test-e2e/KernelFusion/internalize_array_wrapper.cpp:7]: Right. This was my attempt to fix [this error](https://github.com/intel/llvm/actions/runs/5730435105/job/15532799242):

```
FAIL: SYCL :: KernelFusion/internalize_array_wrapper.cpp (1041 of 1567)
******************** TEST 'SYCL :: KernelFusion/internalize_array_wrapper.cpp' FAILED ********************
Script:
--
: 'RUN: at line 2';    /__w/llvm/llvm/toolchain/bin//clang++   -fsycl -fsycl-targets=spir64 /__w/llvm/llvm/llvm/sycl/test-e2e/KernelFusion/internalize_array_wrapper.cpp -fsycl-embed-ir -O2 -o /__w/llvm/llvm/build-e2e/KernelFusion/Output/internalize_array_wrapper.cpp.tmp.out
: 'RUN: at line 3';   env ONEAPI_DEVICE_SELECTOR=ext_oneapi_level_zero:gpu  /__w/llvm/llvm/build-e2e/KernelFusion/Output/internalize_array_wrapper.cpp.tmp.out
: 'RUN: at line 3';   env ONEAPI_DEVICE_SELECTOR=opencl:gpu  /__w/llvm/llvm/build-e2e/KernelFusion/Output/internalize_array_wrapper.cpp.tmp.out
: 'RUN: at line 3';   env ONEAPI_DEVICE_SELECTOR=opencl:cpu  /__w/llvm/llvm/build-e2e/KernelFusion/Output/internalize_array_wrapper.cpp.tmp.out
--
Exit Code: -6

Command Output (stdout):
--
$ ":" "RUN: at line 2"
note: command had no output on stdout or stderr
$ "/__w/llvm/llvm/toolchain/bin//clang++" "-fsycl" "-fsycl-targets=spir64" "/__w/llvm/llvm/llvm/sycl/test-e2e/KernelFusion/internalize_array_wrapper.cpp" "-fsycl-embed-ir" "-O2" "-o" "/__w/llvm/llvm/build-e2e/KernelFusion/Output/internalize_array_wrapper.cpp.tmp.out"
# command stderr:
clang++: warning: argument unused during compilation: '-fsycl-embed-ir' [-Wunused-command-line-argument]

$ ":" "RUN: at line 3"
note: command had no output on stdout or stderr
$ "env" "ONEAPI_DEVICE_SELECTOR=ext_oneapi_level_zero:gpu" "/__w/llvm/llvm/build-e2e/KernelFusion/Output/internalize_array_wrapper.cpp.tmp.out"
# command stderr:
internalize_array_wrapper.cpp.tmp.out: /__w/llvm/llvm/llvm/sycl/test-e2e/KernelFusion/internalize_array_wrapper.cpp:131: int main(): Assertion `all(tmp[id][i][j] == not_written) && "Not internalizing"' failed.

error: command failed with exit status: -6

--
```

**Comment by jsji** [comment]
> @jsji The problem is not really about this test in particular, the problem is basic ESIMD functionality will break for people not using the latest drivers. Even in internal testing we don't seem to be using a driver that has the fix.

In theory, when we release the changes , we can require *NEW* drivers. But yes, we will need to test and update the new drivers in all our relevant tests including internal tests.  And also involve the release managers.

**Comment by aelovikov-intel** [comment]
> @aelovikov-intel, could you run as much tests as possible with the IGC 1.0.14342 or later before the switch? I used to use regular pre-commit, but AFAIK, it doesn't work for testing changes in CI anymore.

Driver update isn't considered to be CI changes. You should be able to test it.

File: sycl/test-e2e/KernelFusion/internalize_array_wrapper.cpp
Comment (sommerlukas) [reply sycl/test-e2e/KernelFusion/internalize_array_wrapper.cpp:7]: Oh, sorry, I misunderstood your previous comment, my bad. I missed that adding the non-opaque flag was your attempt to fix the test. 

I'll approve the changes to the test to deactivate it now.

**Comment by bader** [comment]
> We'll investigate the issue and address in a separate PR.

Great. Thank you!

**Comment by bader** [comment]
@sarnex, the version is use in CI is the latest publicly available IGC version. To update the IGC version, we need to wait for a public release.

> I don't actually know how often these functions are used by customers, maybe @v-klochkov has more customer insight.

@v-klochkov, @turinevgeny, please, approve if it's okay to have `sycl/test-e2e/ESIMD/ext_math.cpp` disabled while we are waiting for a new GPU driver.

**Review by turinevgeny — APPROVED** [review]

**Comment by bader** [comment]
@intel/dpcpp-tools-reviewers, @intel/llvm-reviewers-runtime, ping.

**Review by asudarsa — APPROVED** [review]
Hi @bader 

Overall looks ok to turn this on. I do not see any major issues, though it might help to resolve the failures soon as this will block a few folks. If you think it's a good idea, please open an issue for resolving these fails.

Thanks

**Comment by bader** [comment]
> I do not see any major issues, though it might help to resolve the failures soon as this will block a few folks. 

Who might be blocked by this?

> If you think it's a good idea, please open an issue for resolving these fails.

I already did - https://github.com/intel/llvm/issues/10682.

**Comment by asudarsa** [comment]
> > I do not see any major issues, though it might help to resolve the failures soon as this will block a few folks.
> 
> Who might be blocked by this?
> 

I think bfloat issue might block OneDNN etc. But it was more of a blanket statement. Not a blocker here.
> > If you think it's a good idea, please open an issue for resolving these fails.
> 
> I already did - #10682.

Just spotted it in the earlier comments. Sorry for missing it. I will try to see if I can help there. Thanks again

**Comment by bader** [comment]
@againull, @intel/llvm-reviewers-runtime, ping.

**Review by againull — APPROVED** [review]
No concerns from SYCL RT point of view.

**Pull request merged**
